app/models/assignment.rb

```diff
@@ -634,27 +634,29 @@ def summary_json(user)
 
   # Generates the summary of the most test results associated with an assignment.
   def summary_test_results
-    test_groups_query = self.test_groups
-                            .joins(test_group_results: { test_run: :grouping })
-                            .group('test_groups.id', 'groupings.id')
-                            .select('test_groups.id AS test_groups_id',
-                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
-                            .where.not('test_runs.submission_id': nil)
-                            .to_sql
-
-    self.test_groups
-        .joins(test_group_results: [:test_results, { test_run: { grouping: :group } }])
-        .joins("INNER JOIN (#{test_groups_query}) sub \
-                ON test_groups.id = sub.test_groups_id
-                AND test_group_results_test_groups.created_at = sub.test_group_results_created_at")
+    latest_test_run_by_grouping = TestRun.group('grouping_id').select('MAX(created_at) as test_runs_created_at',
+                                                                      'grouping_id').to_sql
+
+    latest_test_runs = TestRun.joins("INNER JOIN (#{latest_test_run_by_grouping}) latest_test_run_by_grouping \
+                  ON latest_test_run_by_grouping.grouping_id = test_runs.grouping_id \
+                  AND latest_test_run_by_grouping.test_runs_created_at = test_runs.created_at")
+                              .select('id', 'submission_id', 'grouping_id')
+                              .to_sql
+
+    self.test_groups.joins(test_group_results: :test_results)
+        .joins("INNER JOIN (#{latest_test_runs}) latest_test_runs \
+              ON test_group_results.test_run_id = latest_test_runs.id")
+        .joins('INNER JOIN groupings ON latest_test_runs.grouping_id = groupings.id')
```

Comment:
It's a good idea to reduce the number of joins, but in order to take better advantage of ActiveRecord methods, you can move these joins into the definition of `latest_test_runs` (`TestRun.joins(grouping: group)`) and add the relevant fields to the `select`.